include/derecho/core/detail/external_group_impl.hpp

```diff
@@ -0,0 +1,436 @@
+            my_id,
+            getConfUInt32(CONF_DERECHO_P2P_WINDOW_SIZE),
+            view_max_rpc_window_size,
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REPLY_PAYLOAD_SIZE) + sizeof(header),
+            getConfUInt64(CONF_DERECHO_MAX_P2P_REQUEST_PAYLOAD_SIZE) + sizeof(header),
+            view_max_rpc_reply_payload_size + sizeof(header),
+            true,
+            NULL});
+
+    rpc_thread = std::thread(&ExternalGroup<ReplicatedTypes...>::p2p_receive_loop, this);
+}  // namespace derecho
+
+template <typename... ReplicatedTypes>
+ExternalGroup<ReplicatedTypes...>::~ExternalGroup() {
+    thread_shutdown = true;
+    if(rpc_thread.joinable()) {
+        rpc_thread.join();
+    }
+}
+
+template <typename... ReplicatedTypes>
+bool ExternalGroup<ReplicatedTypes...>::get_view(const node_id_t nid) {
+    try {
+        tcp::socket sock = curr_view ? 
+        tcp::socket(std::get<0>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    std::get<PORT_TYPE::GMS>(curr_view->member_ips_and_ports[curr_view->rank_of(nid)]),
+                    false) :
+        tcp::socket(getConfString(CONF_DERECHO_LEADER_IP), getConfUInt16(CONF_DERECHO_LEADER_GMS_PORT));
+        
+        JoinResponse leader_response;
+        bool success;
+        uint64_t leader_version_hashcode;
+        success = sock.exchange(my_version_hashcode, leader_version_hashcode);
+        if(!success) throw derecho_exception("Failed to exchange version hashcodes with the leader! Leader has crashed.");
+        if(leader_version_hashcode != my_version_hashcode) {
+            throw derecho_exception("Unable to connect to Derecho leader because the leader is running on an incompatible platform or used an incompatible compiler.");
+        }
+        success = sock.write(JoinRequest{my_id, true});
+        success = sock.read(leader_response);
+        if(leader_response.code == JoinResponseCode::ID_IN_USE) {
+            dbg_default_error("Error! Leader refused connection because ID {} is already in use!", my_id);
+            dbg_default_flush();
+            throw derecho_exception("Leader rejected join, ID already in use.");
+        }
+        success = sock.write(ExternalClientRequest::GET_VIEW);
+
+        std::size_t size_of_view;
+        success = sock.read(size_of_view);
+        if(!success) {
+            throw derecho_exception("Leader crashed before it could send the initial View! Try joining again at the new leader.");
+        }
+        char buffer[size_of_view];
+        success = sock.read(buffer, size_of_view);
+        if(!success) {
+            throw derecho_exception("Leader crashed before it could send the initial View! Try joining again at the new leader.");
+        }
+        prev_view = std::move(curr_view);
+        curr_view = mutils::from_bytes<View>(nullptr, buffer);
+        return success;
+    } catch(...) {
+        return false;
+    }
```

Comment:
`catch(...)` is generally risky because it throws away all information about errors that have happened, making it harder to handle them properly. Is that really the best way to handle exceptions in this function?

Reply:
I changed it so that inside try block, the function returns false directly when there's a known failure, and only catches `tcp::connection_failure` in case of socket failures.